sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala

```diff
@@ -521,7 +521,7 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
         None
       }
 
-    val schema = userSpecifiedSchema.getOrElse {
+    val schema = userSpecifiedSchema.map(_.asNullable).getOrElse {
```

Comment:
Actually would you mind updating migration guide (https://github.com/apache/spark/blob/master/docs/sql-migration-guide.md)? You can describe, for example, non-nullable schema was not supported properly in previous Spark version so the output schema of `DataFrameReader.json(jsonDataset: Dataset[String])` and `DataFrameReader.csv(csvDataset: Dataset[String])` became nullable which also matches with `DataFrameReader.json(path: String)` and `DataFrameReader.csv(path: String)`.

Reply:
Done. BTW, shall we backport this PR into the previous version?

Reply:
Let's don't port it back. it will be merged into 3.3 only for now.